### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by ZabihullahNooriWardak**
result before solving the n+1 problem : 
285304100-ca7ff061-85c9-4cef-8df9-bca2ce22bd46

result after N+1 problem : 
285304171-e9714ae6-71bd-4f64-9093-e117019d904e

I have created integration tests using Capybara to thoroughly test each view in our project. The tests are organized to ensure a logical flow and cover various aspects of the user interface.
### User Index Page:
Verifies the visibility of usernames, profile pictures, and the number of posts for each user.
Validates the redirection to a user's show page upon clicking a user.
### User Show Page:

1. Ensures the display of the user's profile picture, username, bio, and the first three posts.
2. Includes a button to view all of a user's posts and verifies the redirection to a user's post index page upon clicking a user's post.
### User Post Index Page:
1. Validates the visibility of the user's profile picture, username, post details, comments, likes, and pagination functionality.
2. Verifies the redirection to a post's show page upon clicking a post.
### Post Show Page:
1. Ensures the visibility of the post's title, author, comments, likes, body, and the details of each comment.
2. These tests provide comprehensive coverage of our application's views and user interactions. Your feedback on this pull request is highly appreciated.

### CODE REVIEW

File: spec/integrations/users_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  it 'displays the user\'s recent posts' do
+    visit "users/#{@user.id}"
+    @recent_post.each do |post|
+      expect(page).to have_content(post.title)
+      expect(page).to have_content(post.text)
+    end
+  end
+
```
Comment (Shedrack-Sunday) [inline spec/integrations/users_show_spec.rb:27]: - [ ] You happen to miss a couple of tests needed for this file. 
- I can see the user's username.
 - I can see the number of posts the user has written.
 - When I click to see all posts, it redirects me to the user's post's index page.
 
 - [ ] Kindly add them up 👏 
 
 ### View 👍 
 
image

File: spec/integrations/posts_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  it 'can see the username' do
+    visit path
+    puts user.name
+    find(:css, '.user-photo')
+
+    expect(page).to have_content(user.name)
+  end
+
+  it 'can see number of posts of the user' do
+    visit path
```
Comment (Shedrack-Sunday) [inline spec/integrations/posts_index_spec.rb:32]: - [ ] Kindly add the test `I can see the first comments on a post.`. This particular test is missing 

### View:
image

File: spec/integrations/posts_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  it 'can see the post body' do
+    visit path
+    expect(page).to have_content(post1.text)
+  end
+
+  it 'can see username of each commenter' do
+    visit path
+    expect(page).to have_content(comment1.user.name)
+  end
+end
```
Comment (Shedrack-Sunday) [inline spec/integrations/posts_show_spec.rb:46]: - [ ] Kindly complete the last test, `I can see the comment each comment left.`. Good job 👍 

### View:
image

File: db/migrate/20231026201955_add_index_to_posts.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddIndexToPosts < ActiveRecord::Migration[7.1]
+  def change
+    add_index :posts, :author_id
+  end
+end
```
Comment (Shedrack-Sunday) [inline db/migrate/20231026201955_add_index_to_posts.rb:5]: - [ ] I had issues running this migration due to duplication of the id being passed already in a migration before this. kindly fix this error 👍 

### View:
 
image

**Review by Shedrack-Sunday — CHANGES_REQUESTED** [review]
Hello @ZabihullahNooriWardak     👋 

# CHANGES REQUIRED ♻️

Impressive work on the project so far👏
You've done extremely well so far in this project,  However, there are some adjustments you need to make to move to the next project. 💯



## Highlights   
- Good use of GitHub features ✔️   
- Nice work with Rails testing  features ✔️
- Great work on the  N+1  implementation to fix queries👏   



# - Required Changes  🌵  

- [ ] Kindly note  the comment i made regarding the issues on the error which stems from running your migrations, and a few test cases you missed
# Optional suggestions 

Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing really improves your application.



Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: spec/features/user_show_spec.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      click_on 'Read Post'
+    end
+    expect(page).to have_current_path(post_path(first_post))
+  end
+
+  it 'redirects to the user\'s posts index page when clicking to see all posts' do
+    click_on('See all posts', match: :first)
+    expect(page).to have_current_path(user_posts_path(@user))
+  end
+end
```
Comment (zieeco) [inline spec/features/user_show_spec.rb:59]: - Hey, great job addressing all the changes pointed out by the previous reviewer, however, while you implemented the changes, you made some changes to your file naming which made it tedious to understand the changes you've made. consistent is key to a good code review workflow to foster readability in your codebase. Here is what you've done


   - You renamed the `integrations` to `features` in your spec file, which made it take time for me to trace your changes
   - You also have renamed `spec/integrations/users_show_spec.rb` to `spec/features/user_show_spec.rb`, see the file structure in the image below
 
*spec*  
image

*listing the files and directories in your spec file*
image

---
- Although you've fixed all the required changes, I can't further approve your PR due to the changes you've made to the files and naming. Please, kindly follow the correct file naming convention as in the case of 
 `spec/features/user_show_spec.rb` which is meant to be pluralized `spec/features/users_show_spec.rb`.


- **[OPTIONAL]** Also, if possible rename the `features` to `integrations` as it was before. :muscle: 

Then, you can submit for another review, which your project would be approved.

**Review by zieeco — CHANGES_REQUESTED** [review]
Hi @Team,





Great work on making the changes requested by the previous reviewer 👏 
You've done well implementing some of the changes, however, there are some issues that you still need to work on to go to the next project, but you are almost there!


## To highlight 🎯

- Requested changes implemented ✔️
- Nice design. ✔️
- Descriptive PR. ✔️


## Required Changes ♻️
_Check the comments under the review._




## Optional suggestions
_Check the comments under the review._

**_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._**


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by TedLivist — APPROVED** [review]
## STATUS: APPROVED ✔️ ✔️

Hi @ZabihullahNooriWardak and @mahammad-mostafa,

✔️ Well-written and passing tests

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**